src/rewrite.rs

```diff
@@ -24,10 +26,19 @@ impl<T: Rewrite> Rewrite for ptr::P<T> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct QueryId {
+    pub(crate) shape: Shape,
+    pub(crate) span: SpanData,
+}
+
+pub(crate) type Memoize = Option<HashMap<QueryId, Option<String>>>;
```

Comment:
Is the outer `Option` necessary?

Reply:
Thanks for reviewing! Yes, in case of `None` function should clear the store at the end, but in `Some(empty)` it should fill it and keep it.

Reply:
Hmm wouldn't it be the same to do `clean = map.is_empty()`?

Reply:
```
*map = Some(HashMap::default());
clean = true;
```
If we set `clean = map.is_empty()`, then `*map = Some(HashMap::default());` should become inserting some dummy thing (which isn't good) to map, otherwise subsequent calls will also receive empty map and clean it, while they shouldn't.

Reply:
I see. I wonder if there is a cleaner solution here, such as clearing the map after each `Item` is formatted.